pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/BaseBooleanAggregateFunction.java

```diff
@@ -0,0 +1,252 @@
+
+// TODO: change this to implement BaseSingleInputAggregationFunction<Boolean, Boolean> when we get proper
+// handling of booleans in serialization - today this would fail because ColumnDataType#convert assumes
+// that the boolean is encoded as its stored type (an integer)
+public abstract class BaseBooleanAggregateFunction extends BaseSingleInputAggregationFunction<Integer, Integer> {
+
+  private final BooleanMerge _merger;
+  private final boolean _nullHandlingEnabled;
+
+  protected enum BooleanMerge {
+    AND {
+      @Override
+      int merge(int left, int right) {
+        return left & right;
+      }
+
+      @Override
+      boolean isTerminal(int agg) {
+        return agg == 0;
+      }
+
+      @Override
+      int getDefaultValue() {
+        return 1;
+      }
+    },
+    OR {
+      @Override
+      int merge(int left, int right) {
+        return left | right;
+      }
+
+      @Override
+      boolean isTerminal(int agg) {
+        return agg > 0;
+      }
+
+      @Override
+      int getDefaultValue() {
+        return 0;
+      }
+    };
+
+    abstract int merge(int left, int right);
+
+    abstract boolean isTerminal(int agg);
+
+    abstract int getDefaultValue();
+  }
+
+  protected BaseBooleanAggregateFunction(ExpressionContext expression, boolean nullHandlingEnabled,
+      BooleanMerge merger) {
+    super(expression);
+    _nullHandlingEnabled = nullHandlingEnabled;
+    _merger = merger;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return _nullHandlingEnabled
+        ? new ObjectAggregationResultHolder()
+        : new IntAggregateResultHolder(_merger.getDefaultValue());
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return _nullHandlingEnabled
+        ? new ObjectGroupByResultHolder(initialCapacity, maxCapacity)
+        : new IntGroupByResultHolder(initialCapacity, maxCapacity, _merger.getDefaultValue());
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+
+    if (blockValSet.getValueType() != FieldSpec.DataType.BOOLEAN) {
+      throw new IllegalArgumentException(
+          String.format("Unsupported data type %s for %s", getType().getName(), blockValSet.getValueType()));
+    }
+
+    int[] bools = blockValSet.getIntValuesSV();
+    if (_nullHandlingEnabled) {
+      int agg = getInt(aggregationResultHolder.getResult());
+
+      RoaringBitmap nullBitmap = blockValSet.getNullBitmap();
+      if (nullBitmap == null) {
+        nullBitmap = new RoaringBitmap();
+      } else if (nullBitmap.getCardinality() > length) {
+        return;
+      }
+
+      // early terminate on a per-block level to allow the
+      // loop below to be more tightly optimized (avoid a branch)
+      if (_merger.isTerminal(agg)) {
```

Comment:
(minor) suggest moving this above the nullBitmap check, which is much more costly